### DISCUSSION TITLE
Error recognizing imported files during testing

### DISCUSSION THREAD

**Discussion by joedakwa** [foundry-rs/foundry#4690] [Help]
Hi, I a having issues writing a simple test, quite new to Foundry.

Here is my ```remappings.txt``` file

```
solmate-utils/=lib/solmate/src/utils/
@openzeppelin/=lib/openzeppelin-contracts-upgradeable/
ds-test/=lib/ds-test/src/
```

Here is my foundry.toml

```
[profile.default]
src = 'src'
out = 'out'
libs = ['lib']
remappings = ["openzeppelin/=lib/openzeppelin-contracts/", "openzeppelin/=lib/openzeppelin-contracts-upgradeable/"]
# See more config options https://github.com/foundry-rs/foundry/tree/master/config
```

Here is my test file:

```
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";
import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
import "../src/alloSettings.sol";
import "ds-test/test.sol";


contract alloSettingsTest is Test {

    alloSettings public AlloSettings;

    function setUp() public {
        AlloSettings = new AlloSettings();
        AlloSettings.initialize();
    }

   function testInitialize() public {
    bool success = address(alloSettings).call(abi.encodeWithSignature("initialize()"));
    Assert.isTrue(success, "initialize function call failed");}
}
```

But i am getting an error here:

[⠊] Compiling...
Error: 
Failed to resolve file: "/home/joedakwa/joe/foundry/hello_foundry/lib/openzeppelin-contracts-upgradeable/contracts-upgradeable/access/OwnableUpgradeable.sol": No such file or directory (os error 2).
 Check configured remappings..
    --> "/home/joedakwa/joe/foundry/hello_foundry/src/alloSettings.sol"
        "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol"

Can anyone advise me ?

Thanks

**Comment by mds1** [comment]
If you run `forge remappings` you can see all remappings. Notice how it's looking for a file in: `.../hello_foundry/lib/openzeppelin-contracts-upgradeable/contracts-upgradeable/access/OwnableUpgradeable.sol`

However, the actual path of the file is probably `.../hello_foundry/lib/openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol`

So you'll want to update your remappings so it can find the file

**Reply by joedakwa** [reply]
> If you run `forge remappings` you can see all remappings. Notice how it's looking for a file in: `.../hello_foundry/lib/openzeppelin-contracts-upgradeable/contracts-upgradeable/access/OwnableUpgradeable.sol`
> 
> However, the actual path of the file is probably `.../hello_foundry/lib/openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol`
> 
> So you'll want to update your remappings so it can find the file

So i have this in my current remappsings...

```
ds-test/=lib/solmate/lib/ds-test/src/
forge-std/=lib/forge-std/src/
openzeppelin-contracts-upgradeable/=lib/openzeppelin-contracts-upgradeable/contracts/
openzeppelin-contracts/=lib/openzeppelin-contracts/
solmate/=lib/solmate/src/
```

I would need to change it to this for example?

```
hello_foundry/=lib/openzeppelin-contracts-upgradeable/contracts/
```

**Reply by mds1** [reply]
In your current remappings you have:
```
openzeppelin-contracts-upgradeable/=lib/openzeppelin-contracts-upgradeable/contracts/
openzeppelin-contracts/=lib/openzeppelin-contracts/
```

This part is a problem, as you have the same key of `openzeppelin-contracts` mapping to two different paths, so one will be ignored. To start, I'd suggest deleting all the openzeppelin remappings you have in `remappings.txt` or `foundry.toml`. Then run `forge remappings` and you'll see the defaults that foundry auto-detects. If those are too long and you want shorter ones, then you can add custom remappings. Learn more in the docs here: https://book.getfoundry.sh/projects/dependencies#remapping-dependencies

**Reply by joedakwa** [reply]
> If you run `forge remappings` you can see all remappings. Notice how it's looking for a file in: `.../hello_foundry/lib/openzeppelin-contracts-upgradeable/contracts-upgradeable/access/OwnableUpgradeable.sol`
> 
> 
> 
> However, the actual path of the file is probably `.../hello_foundry/lib/openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol`
> 
> 
> 
> So you'll want to update your remappings so it can find the file

Right I see. It seems to work with another repo where I am using a remappings.txt file, including @openzeppelin/=./node_modules/@openzeppelin/

However, I want to be able to run "forge build" on upgradeable contracts too, which is where I seem to be facing the problem. 

Like this for example: @openzeppelin/=./node_modules/@openzeppelin/contracts-upgradeable doesn't work either?

Should I go ahead with your suggestion?

**Reply by mds1** [reply]
> Should I go ahead with your suggestion?

Yea that would be a good starting point. If you are still having issues, please link to a repo that I can clone to run `forge build` on so it's easy to reproduce

**Reply by joedakwa** [reply]
> > Should I go ahead with your suggestion?
> 
> Yea that would be a good starting point. If you are still having issues, please link to a repo that I can clone to run `forge build` on so it's easy to reproduce

Yea still having issues, here is the link to clone the repo :-)

https://github.com/joedakwa/foundry_test.git

**Reply by mds1** [reply]
This should help https://github.com/joedakwa/foundry_test/pull/1 🙂
